janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java

```diff
@@ -2793,4 +2793,60 @@ public void testOrderByWithRange() {
 
     }
 
+    @Test
+    public void shouldUpdateIndexFieldsAfterIndexModification() throws InterruptedException, ExecutionException {
+        String key1 = "testKey1";
+        String key2 = "testKey2";
+        String key3 = "testKey3";
+        String vertexL = "testVertexLabel";
+        String indexName = "mixed";
+
+        PropertyKey p1 = mgmt.makePropertyKey(key1).dataType(Long.class).make();
+        PropertyKey p2 = mgmt.makePropertyKey(key2).dataType(Long.class).make();
+        mgmt.makeVertexLabel(vertexL).make();
+
+        JanusGraphIndex index = mgmt.buildIndex(indexName, Vertex.class)
+            .indexOnly(mgmt.getVertexLabel(vertexL))
+            .addKey(mgmt.getPropertyKey(key1))
+            .addKey(mgmt.getPropertyKey(key2))
+            .buildMixedIndex(INDEX);
+        if (index.getIndexStatus(p1) == SchemaStatus.INSTALLED) {
+            mgmt.updateIndex(mgmt.getGraphIndex(indexName), SchemaAction.REGISTER_INDEX).get();
+            mgmt.updateIndex(mgmt.getGraphIndex(indexName), SchemaAction.ENABLE_INDEX).get();
+        } else if (index.getIndexStatus(p1) == SchemaStatus.REGISTERED) {
+            mgmt.updateIndex(mgmt.getGraphIndex(indexName), SchemaAction.ENABLE_INDEX).get();
+        }
+        mgmt.commit();
+
+        JanusGraphVertex vertex = graph.addVertex(vertexL);
+        vertex.property(key1, 111L);
+        vertex.property(key2, 222L);
+
+        graph.tx().commit();
+        //By default ES indexes documents each second. By sleeping here we guarantee that documents are indexed.
+        // It is just for testing. A better approach is to use Refresh API.
+        Thread.sleep(1500L);
+
+        List<Vertex> result = graph.traversal().V().hasLabel(vertexL).has(key1, 111L).toList();
+
+        assertEquals(1, result.size());
+
+        mgmt = graph.openManagement();
+        PropertyKey testKey3 = mgmt.makePropertyKey(key3).dataType(Long.class).make();
+        mgmt.addIndexKey(mgmt.getGraphIndex(indexName), testKey3);
+        mgmt.commit();
+
+        ManagementSystem.awaitGraphIndexStatus(graph, indexName).status(SchemaStatus.REGISTERED, SchemaStatus.ENABLED).call();
+
+        mgmt = graph.openManagement();
+        mgmt.updateIndex(mgmt.getGraphIndex(indexName), SchemaAction.REINDEX).get();
+        mgmt.commit();
+
+        vertex = graph.addVertex(vertexL);
+        vertex.property(key1, 1L);
+        vertex.property(key2, 2L);
+        vertex.property(key3, 3L);
+
+        graph.tx().commit();
```

Comment:
Having thought this for a while, I now wonder if we actually should reset the cache of other transactions. Just to recap this bug: an update to the mixed index would cause open transactions to fail if they happen to touch the same index. The fix in this PR tries to make schema changes transparent to all transactions.

Now my question is, what isolation level should we provide? I know JanusGraph barely has any isolation level (some stuff is transparent while others are not), but if we want to "imitate" stronger isolation, say, snapshot isolation, then shouldn't we let those transactions abort? In that case, the fix to the bug would be to provide a better error message.

In fact, while writing test cases, I found out that the fix in this PR is incomplete. If you add 

```java
assertTrue(graph.traversal().V().hasLabel(vertexL).has(key3, 3L).hasNext());
```

after the commit, the test would fail, because the mixed index commit fails (JanusGraph by design does not throw exceptions for mixed index commit). To fix this, we likely need to refresh some other cache as well.

Reply:
> To fix this, we likely need to refresh some other cache as well.

This seems viable but non-trivial. Likely we need to update the information stored in IndexSerializer.IndexInfoRetriever::indexes, if we want to continue in this way.

I am not sure which way we should choose and I am happy to discuss. If we choose to let transactions abort, then we should revert this PR. If we choose to continue on the path suggested by this PR, we need to add a fix (as explained in the previous comment). For the latter case, I still propose to revert this PR before that fix is applied.

Reply:
Good observations! I personally don't have strong opinion on the way we should treat it. I'm OK with both solutions (i.e. abort transactions or provide the fix in IndexSerializer). Nevertheless, I think we should note the behavior in the documentation so that it wouldn't surprise anyone.

Reply:
I don't have a strong opinion either. I have created https://github.com/JanusGraph/janusgraph/pull/2698 which I think is needed anyways. I (temporarily) reverted this PR in #2698 since the fix in this PR is not complete.